rfcs/0042-bitswap.md

```diff
@@ -0,0 +1,263 @@
+    ~protocol_state_proof:t.protocol_state_proof
+    ~staged_ledger_diff:t.staged_ledger_diff
+    ~delta_transition_chain_proof:t.delta_transition_chain_proof
+    ~validation_callback:(Mina_net2.Validation_callback.create_without_expiration ())
+    ()
+in
+Protocol_version.set_current (Protocol_version.create_exn ~major:0 ~minor:0 ~patch:0) ;
+External_transition.Stable.Latest.bin_size_t (conv small_precomputed_block) ;;
+```
+
+### Shipping as a Soft-Fork
+
+We have an option to ship Bitswap as a soft-fork upgrade rather than as a hard-fork upgrade (shoutout to @mrmr1993 for this suggestion; writing this nearly verbatim from him). We can do this by adding a new pub/sub topic for broadcasting the new block header broadcast message format, which we can support in parallel to the current pub/sub topic we use for all of our broadcast messages. We currently use the `"coda/consensus-messages/0.0.1"` topic for blocks, transactions, and snark work. Nodes running the new version of the software can still support this topic, but in addition, can subscribe to and interact over a new topic `"mina/blocks/1.0.0"`, where we can broadcast the new message format for blocks. Nodes are able to filter subscriptions from other nodes based on what they subscribe to, configured using the [`WithSubscriptionFilter` option](https://github.com/libp2p/go-libp2p-pubsub/blob/55d412efa7f5a734d2f926e0c7c948f0ab4def21/subscription_filter.go#L36). Utilizing this, nodes that support the `"mina/blocks/1.0.0"` can filter out the `"coda/consensus-messages/0.0.1"` topic from nodes that support both topics. By filtering the topics like this, nodes running the new version can broadcast new blocks over both topics while avoiding sending the old message format to other nodes which support the new topic. Then, in the next hard fork, we can completely deprecate sending blocks over the old topic.
+
+The main detail to still figure out here is how new nodes will bridge the new topic messages to the old topic. It may be as simple as just broadcasting the block as a fresh message on the old topic, but that isn't normally how a pub/sub rebroadcast cycle would operate. This approach may "just work", but we should do additional research into what differentiates a libp2p pub/sub broadcast from a rebroadcast. If it turns out the only difference is how we select which peers to send the message to, the proposed topic filtering solution may address it automatically.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This adds significant complexity to how the protocol gossips around information. The control flow for validating blocks is more complex than before, and there is new state to synchronize between the processes in the architecture. It also adds new delays to when the full block data will be available to each node (but the tradeoff here is that we are able to more consistently gossip block headers around the network within the same slot those blocks are produced).
```

Comment:
Having the headers alone isn't enough for block producers though, right? They'll need the full new state to be able to produce their blocks.

On this basis, perhaps it's worth mentioning that we'll have to do something with block production so that it doesn't operate against the best tip if it only has the header.

Reply:
Yes, you would need the full block to build the breadcrumb in the transition frontier. Thanks for calling out the added complexity this brings, I should put some more though into how the data flow here should work. I was originally thinking that there would be a step in between initial validation and building a breadcrumb where we wait to download the block body before we actually add it to the transition frontier, rather than adding support in the transition frontier for partial blocks. But maybe this gets too complicated if when we are handling cases where we hear about a block header for a block which is a successor of another block we are still downloading the body for (and thus is not in the frontier yet).